pkg/controller/traceflow/controller.go

```diff
@@ -272,38 +278,27 @@ func (c *Controller) checkTraceflowStatus(tf *opsv1alpha1.Traceflow) (retry bool
 	}
 	if sender && receiver {
 		tf.Status.Phase = opsv1alpha1.Succeeded
-		_, err = c.client.OpsV1alpha1().Traceflows().UpdateStatus(context.TODO(), tf, metav1.UpdateOptions{})
-		return
+		_, err := c.client.OpsV1alpha1().Traceflows().UpdateStatus(context.TODO(), tf, metav1.UpdateOptions{})
+		return err
 	}
 	if time.Now().UTC().Sub(tf.CreationTimestamp.UTC()).Seconds() > timeout {
-		_, err = c.errorTraceflowCRD(tf, "traceflow timeout")
-		return
+		return c.errorTraceflowCRD(tf, "traceflow timeout")
 	}
-	retry = true
-	return
+	return nil
 }
 
-func (c *Controller) runningTraceflowCRD(tf *opsv1alpha1.Traceflow, dataPlaneTag uint8) (*opsv1alpha1.Traceflow, error) {
+func (c *Controller) runningTraceflowCRD(tf *opsv1alpha1.Traceflow, dataPlaneTag uint8) error {
 	tf.Status.DataplaneTag = dataPlaneTag
```

Comment:
Just noticed a potential bug, but not introduced by this PR:
The controller is modifying the object returned by the store's "Get" method, which is highly discouraged: https://github.com/kubernetes/kubernetes/blob/3f579d8971fcce96d6b01b968a46c720f10940b8/staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go#L31-L40.
Although it might be fine for now as we don't build indices on the two fields it's mutating and don't operate a single object concurrently, it's still safer to treat the object as read-only as recommended like other [controllers](https://github.com/kubernetes/kubernetes/blob/3f579d8971fcce96d6b01b968a46c720f10940b8/pkg/controller/endpoint/endpoints_controller.go#L481). It may affect the update event the eventHandler will receive.

Reply:
Sure. Let me change to copy the object.